### PR TITLE
[TIR] Construct the inverse in SuggestIndexMap

### PR DESCRIPTION
**Pull request by vinx13**
Computing the inverse mapping requires arithmetic analysis which is not guaranteed to cover all cases. We provide the pre-defined inverse index map instead.
cc @Hzfengsy @junrushao @junrushao1994

### CODE REVIEW

File: include/tvm/tir/index_map.h [hunk 1]
```diff
@@ -133,16 +145,19 @@ class IndexMapNode : public Object {
...
     v->Visit("initial_indices", &initial_indices);
     v->Visit("final_indices", &final_indices);
+    v->Visit("inverse_index_map", &inverse_index_map);
   }
 
   bool SEqualReduce(const IndexMapNode* other, SEqualReducer equal) const {
     return equal.DefEqual(initial_indices, other->initial_indices) &&
-           equal(final_indices, other->final_indices);
+           equal(final_indices, other->final_indices) &&
+           equal(inverse_index_map, other->inverse_index_map);
```
Comment (junrushao) [inline include/tvm/tir/index_map.h:154]: Do we need to include `inverse_index_map` in structural equality check? Consider a case where this field is optional so `lhs` doesn't have it while `rhs` does, and should they be considered the same?

File: src/tir/schedule/analysis/layout.cc [hunk 2]
```diff
@@ -167,20 +167,25 @@ Optional<IndexMap> SuggestIndexMap(const Buffer& buffer, const Array<PrimExpr>&
     }
     return a.lower_factor > b.lower_factor;
   });
+  // Compute the inverse permutation by argsort
+  std::vector<int> inverse_order = order;
```
Comment (junrushao) [inline src/tir/schedule/analysis/layout.cc:171]: Shall we add a test for `Suggest-Index-Map` where the inverse of the mapping is not detectable?

**Review by junrushao — APPROVED** [review]
LGTM!

File: tests/python/unittest/test_tir_schedule_analysis.py [hunk 3]
```diff
@@ -101,6 +101,53 @@ def test_suggest_index_map_bijective():
...
+    i0_0_i1_0_i2_0_i3_0_i0_1_i1_1_i2_1_i3_1_fused, i3_3_fused, i4_0, i4_1 = _make_vars(
+        "i0_0_i1_0_i2_0_i3_0_i0_1_i1_1_i2_1_i3_1_fused", "i3_3_fused", "i4_0", "i4_1"
+    )
+    eps = floordiv(i0_0_i1_0_i2_0_i3_0_i0_1_i1_1_i2_1_i3_1_fused, 336) * 2 + floordiv(
+        floormod(i0_0_i1_0_i2_0_i3_0_i0_1_i1_1_i2_1_i3_1_fused, 16), 8
+    )
+    nu = floordiv(floormod(i0_0_i1_0_i2_0_i3_0_i0_1_i1_1_i2_1_i3_1_fused, 336), 112) * 2 + floordiv(
+        floormod(i0_0_i1_0_i2_0_i3_0_i0_1_i1_1_i2_1_i3_1_fused, 8), 4
+    )
+    co = floormod(i0_0_i1_0_i2_0_i3_0_i0_1_i1_1_i2_1_i3_1_fused, 4) * 32 + i3_3_fused
```
Comment (junrushao) [inline tests/python/unittest/test_tir_schedule_analysis.py:115]: 😂 shall we somehow use a less cryptic name?

File: src/tir/schedule/analysis/layout.cc [hunk 4]
```diff
@@ -167,20 +167,25 @@ Optional<IndexMap> SuggestIndexMap(const Buffer& buffer, const Array<PrimExpr>&
...
+  std::sort(inverse_order.begin(), inverse_order.end(),
+            [&order](int _a, int _b) -> bool { return order[_a] < order[_b]; });
   // Step 5. Create the indexing mapping
   auto f_alter_layout = [f_flatten_index = std::move(f_flatten_index),  //
-                         split_exprs = std::move(split_exprs),          //
-                         order = std::move(order),                      //
-                         shape = buffer->shape,                         //
+                         &split_exprs,                                  //
+                         &order,                                        //
+                             & shape = buffer->shape,                   //
```
Comment (Hzfengsy) [inline src/tir/schedule/analysis/layout.cc:178]:
```suggestion
                        & shape = buffer->shape,                        //
```

**Review by Hzfengsy — APPROVED** [review]

File: src/tir/schedule/analysis/layout.cc (hunk 4 above)
Comment (vinx13) [reply src/tir/schedule/analysis/layout.cc:178]: clang format somehow prefers this way

**Pull request merged**